wcfsetup/install/files/lib/system/trophy/condition/TrophyConditionHandler.class.php

```diff
@@ -106,4 +130,36 @@ private function getUserIDs(Trophy $trophy) {
 		
 		return $userList->getObjectIDs();
 	}
+	
+	/**
+	 * Returns the userTrophyIDs of the users, which no longer fulfills the trophy conditions. 
+	 * 
+	 * @param       Trophy          $trophy
+	 * @return      integer[]
+	 */
+	private function getRevocableUserTrophyIDs(Trophy $trophy) {
```

Comment:
I'm not sure how the generated query looks like. Can you show one together with the `EXPLAIN` for some example data? I feel like it can be simplified, if one looks at the information required:

1. We need the set of users that currently have the trophy assigned.
2. We need the set of users that should have the trophy assigned.
3. We determine the difference of set 1 and set 2: The set of users that have the trophy assigned, but should not have it.

You should be able to achieve this with a `RIGHT JOIN`, while still being able to use the usual processing:

1. Create the UserList, attach all the conditions for the trophy to be awarded
2. Add an sqlConditionJoin, `RIGHT JOIN`ing the trophy table `ON user.userID = trophy.userID`. This creates a row for every row in the trophy table, no matter whether a row in the user table exists.
3. Adding a `WHERE trophy.trophyID = $trophy->trophyID AND user.userID IS NULL`. This filters the rows to receive only rows matching the given trophy, where *no* row in the user table exists. Which is exactly all assigned trophies, without the trophies with should be revoked.
4. `SELECT trophy.trophyID`.

The resulting query should look something like this:

```
SELECT     trophy.userTrophyID
FROM       wcf1_user              u
LEFT JOIN  wcf1_user_option_value uov ON u.userID = uov.userID
RIGHT JOIN wcf1_user_trophy       ut  ON u.userID = ut.userID
WHERE ($assignmentConditions)
  AND ut.trophyID = $trophyID
  AND u.userID    IS NULL
LIMIT 500;
```